src/rest_framework_dso/paginator.py

```diff
@@ -0,0 +1,241 @@
+        returned by the base class.
+        """
+        iterator = super().iterator(*args, **kwargs)
+        return self._set_observable_iterator(iterator)
+
+    def __iter__(self):
+        """Return observable iterator and add observer.
+        Wraps an observable iterator around the iterator
+        returned by the base class.
+        """
+        return self._set_observable_iterator(super().__iter__())
+
+    def _set_observable_iterator(self, iterator: Iterable) -> ObservableIterator:
+        """Wrap an iterator inside an ObservableIterator"""
+        iterator = ObservableIterator(iterator)
+        iterator.add_observer(self._item_observer)
+
+        # Remove observer from existing oberservable iterator
+        if self._obs_iterator is not None:
+            self._obs_iterator.clear_observers()
+
+        self._obs_iterator = iterator
+
+        # Notify observers of empty iterator
+        if not iterator:
+            self._item_observer(None, True)
+
+        return iterator
+
+    def add_observer(self, callback: Callable):
+        """Install an observer callback that is notified when items are iterated"""
+        self._item_callbacks.append(callback)
+
+    def _item_observer(self, item, is_empty=False):
+        """Notify all observers."""
+        for callback in self._item_callbacks:
+            callback(item, self._obs_iterator, is_empty)
+
+    def is_iterated(self) -> bool:
+        """The iterator has finished"""
+        return self._obs_iterator is not None and self._obs_iterator._is_iterated
+
+
+class DSOPage(DjangoPage):
+    """A page that can be streamed.
+
+    This page avoids count queries by delaying calculation of the page length.
+
+    The number of items on the page cannot be known before
+    the object_list has been iterated.
+    Therefore __len__ and has_next methods are only valid
+    after the object list has been iterated.
+    """
+
+    def __init__(self, object_list, number, paginator):
+        self.number = number
+        self.paginator = paginator
+        self._length = 0
+        self._has_next = None
+        if isinstance(object_list, QuerySet):
+            # We have to cast the queryset instance into an observable queryset here. Not pretty.
+            # Pagination in DRF is handled early on in the pipeline where, because we stream the
+            # data we don't know the number of items on the page.
+            # We need the number of items to tell wether a next page exists.
+            #
+            # So we need to keep track of the iteration process happening in the renderer.
+            # We can not create the iterator at this point, because streaming will break
+            # further down the line if object_list is not a queryset
+            # And we dont want to move all the pagination logic into the generic renderer.
+            # So in order to watch the iterators created later on by the queryset we have to
+            # wrap it here.
```

Comment:
👏 perfect! heel fijn dat dit de pijnpunten en overweging van deze "Not pretty" situatie uitlegd.